### PR TITLE
[MNG-8001] Expose CoreRealm internally

### PR DESCRIPTION
**Pull request by cstamas**
Expose internally (not for plugin API) CoreRealm.

---

https://issues.apache.org/jira/browse/MNG-8001

### CODE REVIEW

File: api/maven-api-core/src/main/java/org/apache/maven/api/services/RealmService.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+import org.codehaus.plexus.classworlds.ClassWorld;
+import org.codehaus.plexus.classworlds.realm.ClassRealm;
+
+/**
+ * Access to {@link ClassRealm}.
+ *
+ * @since 4.0.0
+ */
+@Experimental
+public interface RealmService extends Service {
```
Comment (rmannibucau) [inline api/maven-api-core/src/main/java/org/apache/maven/api/services/RealmService.java:33]: not sure the goal but plexus should probably not surface in the API so maybe just use `ClassLoader`?

Comment (cstamas) [reply api/maven-api-core/src/main/java/org/apache/maven/api/services/RealmService.java:33]: Hm, you are right. Ultimate goal is "contain" scattered use of PlexusContainer. So let me change this PR, and offer single method w/ Classloader as you propose, but add "internal service" (usable in maven core only) that delivers what this PR currently has (as this is needed for 2 components that mis-used PlexusContainer only to obtain what they needed)

File: maven-core/src/main/java/org/apache/maven/internal/CoreRealm.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import org.codehaus.plexus.classworlds.ClassWorld;
+import org.codehaus.plexus.classworlds.realm.ClassRealm;
+
+/**
+ * Access to core {@link ClassRealm}.
+ *
+ * @since 4.0.0
+ */
+@Experimental
+public interface CoreRealm {
```
Comment (gnodet) [inline maven-core/src/main/java/org/apache/maven/internal/CoreRealm.java:32]: The name sounds weird given we have `getCoreRealm()` as a method...

Comment (cstamas) [reply maven-core/src/main/java/org/apache/maven/internal/CoreRealm.java:32]: Method renamed to `getRealm()`, so it is now `CoreRealm.getRealm()`

**Comment by michael-o** [comment]
Why is this necessary?

**Comment by cstamas** [comment]
> Why is this necessary?

To contain passing around PlexusContainer, basically to remove it (or "hide" it well) even in Maven core

**Comment by michael-o** [comment]
> > Why is this necessary?
> 
> To confine passing around PlexusContainer, basically to remove it (or "hide" it well) even in Maven core

OK, this basically applies a decorator here.

**Review by gnodet — APPROVED** [review]

**Comment by jira-importer** [comment]
Resolve #10669

**Pull request merged**